### ISSUE TITLE
To do as of November 13 2018

### ISSUE THREAD

**Issue by fdreyer** [JetsGame/GroomRL#1]
Step 1
* [x]  Reward function: best width and best functional form
* [x]  Reward function: add a Soft drop component?
  * One could try to reward both grooming of soft wide-angle radiation, and non-grooming of hard collinear radiation. E.g. 
     reward_SD = N * [ exp(-ln(Delta) ln(z)) Theta(groom==true) + (1-exp(ln(Delta)*ln(z))) Theta(groom==false)], where N is some normalisation factor < 1 controlling how much this term of the reward enters into the full reward function.  Possibly one could also add some additional exponents/factors to allow for more tuning
![image](/uploads/dd196ba721a3b345c1fa34d7972177d8/image.png)
* [x]  Random grid search of hyperparameters
  * Optimizer
  * Architecture (# layers, LSTM/Dense/...)
  * Training validation, k-folding
* [x]  Restructure the internal bookkeeping to allow for more flexible analysis => more general observable outputs
  * [x] Groom also along the two subjet momenta of a declustering, not just the main node. This requires better tracking of IDs: perhaps use ID = -1 for subjets with no constituents, then order the children IDs in pt: children=[ID_j1, ID_j2]
  * [x] Output the groomed jet constituents instead of only the mass when running in test mode. (This requires removing the momentum of groomed radiation also along the two subjets.)
  * [x] Write some routines that read in the list of groomed jet constituents, and returns observables for each jet
* [x] Think about way of exporting the groomer without the gym and keras-rl libraries
  * For that we need to understand how the DQN takes a decision based on the NN output after seeing a given state. Is it just a threshold on the output node? If so it should be easy to use only the NN without going through the DQN, and groom events separately from the DQN environment.
* [x]  What data to use for training: C/A v anti-kt jets, pt cut, W v top v QCD, which MC generator, radius, mixed samples?

Step 2
* [x]  Add additional input observables like mass, azimuthal angle, kt, ....
* [x]  Try a Double DQN (https://arxiv.org/pdf/1509.06461.pdf) setup (it's an option of DQN so should be easy) 
* [x]  Comparisons with Soft Drop, RSD, Trimming, Filtering, ....
* [x]  Visualization tools, Lund plane?
* [ ]  make a video game 

Step 3
* [ ]  Apply to data unrelated to training, like top, pileup
* [x]  Resilience to NP effects, detector effects (maybe do a resilience v efficiency plot?)
